### PR TITLE
Use Airflow base Docker image

### LINKED ISSUE #532: Investigate using Airflow's docker image as a base
## Description

We're presently using `python:3.10-slim` as our base image. Airflow provides their own images that we could potentially use as an alternative base: https://airflow.apache.org/docs/docker-stack/index.html.

We should investigate whether this would be advantageous for us (easier to upgrade versions, not have to manage constraints, smaller image size, etc).

## Additional context

See #502 for more context

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

### PR DESCRIPTION
**Pull request by AetherUnbound**
## Fixes

Fixes #532 by @AetherUnbound

## Description

This PR replaces our current custom Airflow docker image with the official upstream Airflow docker image. With this comes a number of complications and changes. I don't see this as a solid "win". There are lots of pros and cons that are worth weighing before we decide to merge (or not) this change:

**Pros**:
- Running database migrations on startup is baked into the image for us and just requires setting an environment variable.
- Creating a web UI admin user is as easy as setting a few variables, we no longer need to add logic for this.
- We can remove the bespoke `wait_for_db.py` script since [waiting for the database to be ready is baked in](https://airflow.apache.org/docs/docker-stack/entrypoint.html#waits-for-airflow-db-connection).
- The main processes no longer runs as the `root` user, which [is more secure](https://docs.bitnami.com/tutorials/why-non-root-containers-are-important-for-security) and follows the principle of least privilege.
- Reduced maintenance burden - the Airflow folks put in a lot of effort to make sure that the Docker images are production ready.
- We would have separate containers for the webserver & scheduler (rather than having them run on the same container), which helps isolate those services more.

**Cons**:
- Running as a non-root user can cause complications with bind-mounted volumes. Permissions on the host machine will no longer match up with the Docker user, so certain operations have to be adjusted. This can be seen in changes to the `justfile` (specifically the `generate-dag-docs` recipe, which can no longer write out the file without special considerations) or to tests (see the changes to `test_create_provider_ingester.py`). This is not necessarily a bad thing, but it has brought me a bit of frustration and can cause trouble and strange behavior down the line.
- Startup time is slower. Since we're now handing over management of the migrations to Airflow and the upstream entrypoint has [lots of other startup behavior](https://airflow.apache.org/docs/docker-stack/entrypoint.html), there's more that runs at each container start. In production this is not a problem (since the containers shouldn't be starting & stopping regularly), but it is noticeable in development. Many of our `just` recipes take a few seconds longer to run. Nothing insurmountable from my perspective, just something I've noticed.
- Held back on Python releases. We were able to jump on Python 3.10 before it was officially supported by Airflow. This was because Airflow has a large number of provider libraries which it also supports, and it needed to ensure compatibility with all of those libraries before the application as a whole was ready for Python 3.10. We use a very small portion of the available providers, so as it stands we are held back from upgrading to 3.11 until those upstream issues with providers we don't use are resolved (see #855).
- We now need to set up a CSRF secret key for **both development and production environments** (they are currently defined in neither).

**Other changes**:
- Airflow home folder got changed from `/usr/local/airflow` to `/opt airflow`
- I fixed an issue with the create provider ingester tests wherein we were accidentally handing in a string where a list should go and it was producing one provider for each letter of the word `image`.

## Testing Instructions

Run `just build` on this branch. Make sure you're able to run every `just` recipe we have without error. You may need to run `just recreate` if you encounter any errors initially.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by AetherUnbound** [comment]
As expected, there are a few issues with permissions now that we're using a non-root docker user. One test and the `generate-dag-docs` just recipe fails. I'll try to think about how to get around this.

File: docker/airflow/Dockerfile
Comment (sarayourfriend) [inline docker/airflow/Dockerfile:38]: Why no more `upgrade`?

Comment (AetherUnbound) [reply docker/airflow/Dockerfile:38]: There were some serious version conflicts/difficulties in doing this with the Airflow image. I should have noted exactly what the problem was, I can try and re-add it to note the specific error I received.

File: docker-compose.override.yml
Comment (AetherUnbound) [reply docker-compose.override.yml:122]: This is for retaining `ipython` history across docker restarts. Previously we were mounting a local folder to accomplish this (which I'm not even sure this actually worked as expected due to permissions issues) but now that we're running as the `airflow` user within the container it's a LOT more difficult to try and manage mounting that and making it write-accessible to the `airflow` user. It's way easier to simply add a volume mount here instead, since docker manages those entirely and permissions will be preserved. The downside there is that `just recreate` will destroy that history, but I still find it useful to retain the history where possible during development!

**Comment by zackkrida** [comment]
Weighing the pros and cons I think I am still in favor of this work. The main con I'm actually concerned with is "Running as a non-root user can cause complications with bind-mounted volumes." because it admittedly sounds a bit vague. Is there anything we can do in terms of documentation to help mitigate future pain from this one? The others feel like reasonable tradeoffs but this one and it's ability to cause future unforeseen frustration or issues seems unfortunate.

**Comment by rwidom** [comment]
> Weighing the pros and cons I think I am still in favor of this work. The main con I'm actually concerned with is "Running as a non-root user can cause complications with bind-mounted volumes." because it admittedly sounds a bit vague. Is there anything we can do in terms of documentation to help mitigate future pain from this one? The others feel like reasonable tradeoffs but this one and it's ability to cause future unforeseen frustration or issues seems unfortunate.

Yeah, does this mean that code changes will no longer be automatically synced to the docker container? That could mean a lot more `just recreate`s for developers. Or maybe I'm misunderstanding and/or there is a workaround I'm not thinking of... ???

**Comment by AetherUnbound** [comment]
> Weighing the pros and cons I think I am still in favor of this work. The main con I'm actually concerned with is "Running as a non-root user can cause complications with bind-mounted volumes." because it admittedly sounds a bit vague. Is there anything we can do in terms of documentation to help mitigate future pain from this one?

I think a lot of the concern I had in that department was centered around 1) the [issues I had in a personal project](https://github.com/OrcaCollective/techbloc-airflow/pull/8) which required [mounting SSH keys in a peculiar way to ensure permissions were preserved](https://github.com/OrcaCollective/techbloc-airflow/pull/7/files#diff-6f9d41d046756f0ddc2fcee0626bdb50100d12b88f293734eff742818e03efa2R64-R73) and 2) the [workaround we have to do for the DAG doc generation script](https://github.com/WordPress/openverse-catalog/pull/874/files#diff-deb9bb56fb122db0b605aa5b63f95a4665c905b18dd670e1fa6c877576a94ff1R126) and [provider script generation tests](https://github.com/WordPress/openverse-catalog/pull/874/files#diff-e8749752ee7270716e940381f68cfd05deaacfcef29423e7f1e82a0fcf9dcd19R51-R57). Admittedly I was able to find a solution for the former (and most of the trouble I had with it was an unrelated issue with `cp dir/*` not receiving any contents and failing because of _that_, not permissions).

This issue will only ever come up in the following scenarios:
1. We need to add a new bind mount to the container which requires write permissions
2. We add a new script/task/something that needs to write to an _existing_ bind mount (a la DAG doc generation or provider script generation)

Given all that context, where do you think documentation for this scenario might be best suited?

> Yeah, does this mean that code changes will no longer be automatically synced to the docker container? That could mean a lot more just recreates for developers. Or maybe I'm misunderstanding and/or there is a workaround I'm not thinking of... ???

Good clarification question - this should only affect _write permissions_ for bind-mounted volumes. Places where we only need to read files (like for DAG parsing!) should not be affected! Auto-reload and (in production) auto-deployment will both work as they did before 😄

**Comment by stacimc** [comment]
This looks good to me. `just` recipes are running well -- I'm actually not even seeing a perceptible difference in how long it takes to run them locally. Initially had CSRF problems while playing around with local Airflow, but all is well after I remembered to update the `.env` with the secret key 🙂 (including testing auto-reload!)

Looking at pros/cons, I'm happy to move forward with this. Just echoing a (small) concern about documentation given the need for special handling in this scenario:

> We add a new script/task/something that needs to write to an existing bind mount (a la DAG doc generation or provider script generation, more likely but still very uncommon)

Maybe we could throw something into the `docs` folder? Or even just a small note in the main README 🤔

**Comment by AetherUnbound** [comment]
Thanks Staci! I was trying to think where the best place for a piece of info like that to go would be (`justfile` since it's likely where we'd add the script? `docker-compose.*.yml` since that'd be where we might add a bind mount if needed?), README seems like the best and most general place 👍🏼

**Comment by AetherUnbound** [comment]
Just a note for myself: **I will need to update the Airflow version** (and this can supersede #911)

**Comment by AetherUnbound** [comment]
I'm actually going to hold off on upgrading to Airflow 2.5.0 in this PR since the changes are non-trivial, but I've made #924 to track that.

**Comment by AetherUnbound** [comment]
@krysal @stacimc this should be ready for review now!

**Comment by AetherUnbound** [comment]
I've tested this as well by building & starting the image from `main`, then changing my local repo to this PR. DAGs still function and nothing should break 😄

**Comment by stacimc** [comment]
After `just build` on main, switching to this branch, and running `just build` once more, everything looks good in my local Airflow. Old DagRuns are still present, and I was able to run several DAGs as tests (tested with `jamendo`, `metropolitan`, and `audio_data_refresh`).

Locally I'm seeing three test failures consistently with `PermissionError`s, though:

```
         - ../../usr/local/airflow/tests/dags/common/storage/test_audio.py:103 test_AudioStore_add_item_flushes_buffer
         - tests/dags/common/storage/test_media.py:91 test_MediaStore_add_item_flushes_buffer
         - ../../usr/local/airflow/tests/dags/providers/provider_api_scripts/test_cleveland_museum.py:168 test_handle_response
```

**Comment by AetherUnbound** [comment]
@stacimc you may need to run `just recreate` to recreate the volumes those tests write too, since they were likely created with the old `root` user and the new `airflow` user doesn't have access.

**Review by stacimc — APPROVED** [review]
Passing after a recreate 😃 LGTM

File: docker/airflow/Dockerfile
Comment (krysal) [inline docker/airflow/Dockerfile:3]: Minor suggestion for clarity:
```suggestion
ARG PROJECT_PYTHON_VERSION
```

File: docker/airflow/Dockerfile
Comment (krysal) [inline docker/airflow/Dockerfile:44]: New lines helps the understanding of each statement:
```suggestion

RUN mkdir -p ${OUTPUT_DIR} /home/airflow/.ipython && \
    chown airflow ${OUTPUT_DIR} /home/airflow/.ipython
```
and maybe add a comment that is for retaining `ipython` history?

**Review by krysal — APPROVED** [review]
I'd like to deep dive more into what the base Airflow image brings but don't want to withhold this any longer. So approving to unblock it ✅

I left some minor suggestions.

File: docker/airflow/Dockerfile
Comment (AetherUnbound) [reply docker/airflow/Dockerfile:3]: We'd need to change this in a few places (I think in the CI/CD files files as well):

https://github.com/WordPress/openverse-catalog/blob/fba7566540343549b5a01211e69c9237c0ed6863/.github/workflows/ci_cd.yml#L83

If it's okay with you, I'll defer this for now and we can make the change in a separate PR 🙂

File: docker/airflow/Dockerfile
Comment (AetherUnbound) [reply docker/airflow/Dockerfile]: Done in 31988874c2ae6f7d9062d80a47c1d4ee341f5b0f!

**Pull request merged**
